### PR TITLE
[WIP] [DO NOT TOUCH] Fix Magneto to PHP 8 compatibility [CHECK PHP8]

### PR DESCRIPTION
**Pull request by karyna-t**
### Description (*)

This pull request provides the update of the PHP version in require section to "~7.4.0||~8.0.0||~8.1.0" into all composer files.
Also, this PR provides the update composer dependencies.
Finally, this PR contains fixes for Magento to be able to launch with PHP 8.0
Builds of this PR should run with PHP 8.0 environment.

### Related Pull Requests
https://github.com/magento/partners-magento2-infrastructure/pull/63

https://github.com/magento/partners-magento2ee/pull/603
https://github.com/magento/partners-magento2b2b/pull/588
https://github.com/magento/magento2-sample-data/pull/171
https://github.com/magento/inventory/pull/3324
https://github.com/magento/adobe-stock-integration/pull/1868
https://github.com/magento/security-package/pull/306
https://github.com/magento/magento2-page-builder/pull/776
https://github.com/magento/magento2-page-builder-ee/pull/192
https://github.com/magento/adobe-ims/pull/12

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by andrewbess** [comment]
@magento run Unit Tests with env PHP 8.0

**Comment by andrewbess** [comment]
@magento run Unit Tests with env PHP 8.0

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php [hunk 1]
```diff
@@ -207,7 +206,7 @@ protected function scan()
                     goto SCANNER_NAMESPACE_CONTINUE;
                 }
 
-                if ($this->tokenType === T_NS_SEPARATOR || $this->tokenType === T_STRING) {
+                if (\in_array($this->tokenType, $namespaceContentTokenTypes, true)) {
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php]: Could you replace it with a map and looking for a key in this map? 
It works way faster, especially performance is critical for DI file scanner

File: setup/src/Magento/Setup/Module/Di/Code/Scanner/PhpScanner.php [hunk 2]
```diff
@@ -197,16 +197,31 @@ public function collectEntities(array $files)
...
+                    break;
+                }
+            }
+            return join('', $namespaceParts);
+        }
+
+        $namespace = '';
+        for ($tokenOffset = ($tokenIterator + 1); $tokenOffset < $count; ++$tokenOffset) {
+            if ($tokens[$tokenOffset][0] === T_NAME_QUALIFIED
+                || $tokens[$tokenOffset][0] === T_NAME_FULLY_QUALIFIED
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Module/Di/Code/Scanner/PhpScanner.php]: Looks like there should be different cases for T_NAME_QUALIFIED and T_NAME_FULLY_QUALIFIED.
Example: 
https://github.com/Codeception/Codeception/pull/6000/files#diff-7c1158ad3505bce7e9cd0901e6d103a4a6eb81e7f6739c520b28df7d0ed4feb3R149

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php [hunk 3]
```diff
@@ -264,7 +263,7 @@ protected function scan()
                         goto SCANNER_USE_CONTINUE;
                     }
 
-                    if ($this->tokenType == T_NS_SEPARATOR || $this->tokenType == T_STRING) {
+                    if (\in_array($this->tokenType, $namespaceContentTokenTypes, true)) {
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php]: Could you replace it with a map and looking for a key in this map?
It works way faster, especially performance is critical for DI file scanner

File: lib/internal/Magento/Framework/Code/Reader/NamespaceResolver.php [hunk 4]
```diff
@@ -104,7 +105,13 @@ public function getImportedNamespaces(array $fileContent)
...
                         $import,
                         function ($token) {
-                            $whitelist = [T_NS_SEPARATOR, T_STRING, T_AS];
+                            $whitelist = [
+                                T_NS_SEPARATOR,
+                                T_STRING,
+                                T_AS,
+                                T_NAME_QUALIFIED,
+                                T_NAME_FULLY_QUALIFIED
+                            ];
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Code/Reader/NamespaceResolver.php]: please replace to map

**Comment by andrewbess** [comment]
@magento give me test instance

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php (hunk 1 above)
Comment (karyna-t) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileScanner.php]: thanks for the tip, done

File: setup/src/Magento/Setup/Module/Di/Code/Scanner/PhpScanner.php (hunk 2 above)
Comment (karyna-t) [reply setup/src/Magento/Setup/Module/Di/Code/Scanner/PhpScanner.php]: I've double-checked this piece of code and T_NAME_FULLY_QUALIFIED is not needed here, as in this particular place there is no situation when fully qualified namespace can occur. so I removed it from the condition

**Comment by ihor-sviziev** [comment]
Maybe after recent changes?
@magento run Unit Tests with env PHP 8.0

**Comment by andrewbess** [comment]
@magento run Integration Tests with env PHP 8.0

**Comment by andrewbess** [comment]
@magento run Integration Tests with env PHP 8.0

**Comment by andrewbess** [comment]
@magento run Integration Tests with env PHP 8.0

**Comment by andrewbess** [comment]
@magento run Integration Tests with env PHP 8.0

**Comment by andrewbess** [comment]
@magento run WebAPI Tests with env PHP 8.0

**Comment by andrewbess** [comment]
@magento run Integration Tests with env PHP 8.0

**Comment by anzin** [comment]
@magento run Integration Tests with env PHP 8.0

**Comment by eduard13** [comment]
@magento run Integration Tests with env PHP 8.0

**Comment by eduard13** [comment]
@magento run Integration Tests with env PHP 8.0 with extensions magento-commerce/security-package-b2b:platform-health

**Comment by eduard13** [comment]
@magento run Integration Tests with env PHP 8.0 with extensions magento-commerce/security-package-b2b:platform-health

**Pull request closed without merging**